src/main/java/org/gaul/s3proxy/CrossOriginResourceSharing.java

```diff
@@ -0,0 +1,139 @@
+        this.allowedHeaders = new ArrayList<String>();
+
+        if (!Strings.isNullOrEmpty(allowedOrigins)) {
+            for (String origin: allowedOrigins.split(VALUE_SEPARATOR)) {
+                this.allowedOrigins.add(
+                        Pattern.compile(origin, Pattern.CASE_INSENSITIVE));
+            }
+        }
+        if (!Strings.isNullOrEmpty(allowedMethods)) {
+            this.allowedMethods.addAll(Arrays.asList(allowedMethods.split(
+                    VALUE_SEPARATOR)));
+        }
+        if (!Strings.isNullOrEmpty(allowedMethods)) {
+            this.allowedHeaders.addAll(Arrays.asList(allowedHeaders.split(
+                    VALUE_SEPARATOR)));
+        }
+
+        this.allowedMethodsRaw = Joiner.on(HEADER_VALUE_SEPARATOR).join(
+                this.allowedMethods);
+        this.allowedHeadersRaw = allowedHeaders;
+    }
+
+    public String getAllowedMethods() {
+        return this.allowedMethodsRaw;
+    }
+
+    public boolean isOriginAllowed(String origin) {
+        if (!Strings.isNullOrEmpty(origin)) {
+            for (Pattern pattern: this.allowedOrigins) {
+                Matcher matcher = pattern.matcher(origin);
+                if (matcher.matches()) {
+                    logger.debug("CORS origin allowed: {}", origin);
+                    return true;
+                }
+            }
+        }
+        logger.debug("CORS origin not allowed: {}", origin);
+        return false;
+    }
+
+    public boolean isMethodAllowed(String method) {
+        if (!Strings.isNullOrEmpty(method)) {
+            if (this.allowedMethods.contains(method)) {
+                logger.debug("CORS method allowed: {}", method);
+                return true;
+            }
+        }
+        logger.debug("CORS method not allowed: {}", method);
+        return false;
+    }
+
+    public boolean isEveryHeaderAllowed(String headers) {
+        if (!this.allowedHeadersRaw.equals(ALLOW_ANY_HEADER)) {
+            for (String header: headers.split(HEADER_VALUE_SEPARATOR)) {
+                if (!this.allowedHeaders.contains(header)) {
+                    logger.debug("CORS headers not allowed: {}", headers);
+                    return false;
+                }
+            }
+        }
+        logger.debug("CORS headers allowed: {}", headers);
+        return true;
+    }
+
+    @Override
+    public boolean equals(Object object) {
+        if (this == object) {
+            return true;
+        }
+        if (object == null || getClass() != object.getClass()) {
+            return false;
+        }
+
+        CrossOriginResourceSharing that = (CrossOriginResourceSharing) object;
+        return this.allowedOrigins.equals(that.allowedOrigins) &&
+                this.allowedMethodsRaw.equals(that.allowedMethodsRaw) &&
+                this.allowedHeadersRaw.equals(that.allowedHeadersRaw);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.allowedOrigins.hashCode(),
```

Comment:
Do you want to hash `allowedOrigins.hashCode` or just `allowedOrigin`?  The latter is more consistent.